### ISSUE TITLE
Quorum based distribution fails when calling `choose_node` on `start_child`

### ISSUE THREAD

**Issue by Ivor** [derekkraan/horde#49]
The `UniformQuorumDistribution` strategy, and any similar ones where `has_quorum?/1` can return `false` and `choose_node` therefore `nil` breaks the `:start_child` call in `SupervisorImpl`.

```
defmodule Horde.UniformQuorumDistribution do
  def choose_node(identifier, members) do
    if has_quorum?(members) do
      Horde.UniformDistribution.choose_node(identifier, members)
    else
      nil
    end
  end
```
and
```
defmodule Horde.SupervisorImpl do
  def handle_call({:start_child, child_spec} = msg, from, %{node_id: this_node_id} = state) do
    case state.distribution_strategy.choose_node(child_spec.id, state.members) do
      {^this_node_id, _} ->
        {reply, new_state} = add_child(child_spec, state)
        {:reply, reply, new_state}

      {other_node_id, _} ->
        proxy_to_node(other_node_id, msg, from, state)
    end
  end
```

If a cluster is still starting up there will always be a point at which `state.members` < whatever constitutes quorum.

Is there perhaps a way to delay the starting of child processes until the cluster `has_quorum?`?

What would be the best way to deal with the startup and quorum?

**Comment by derekkraan** [comment]
Thanks again for the bug report. We can't guarantee that at any point in time, the cluster has quorum. I think the best thing we can do here is add `nil` to the case statement in `handle_call({:start_child, ...})` so that we gracefully handle the case where there is no quorum.

**Comment by Ivor** [comment]
It seems that there is a big difference between lack of quorum because we're still starting up and lack of quorum because we previously had quorum and we've lost some nodes. 

If the starting of processes was something that only started once we atleast reached quorum for the first time then we can prevent this problem. In other words, have the horde start up and hang on witg starting any processes. Monitor the horde in a loop until we have established quorum and then start the processes. 

Is this something you think is feasible?

**Comment by Ivor** [comment]
Alternatively, if we can store the fact that at some earlier time we did have quorum we can consider a failure of quorum to be a legitimate problem, while if we have never had quorum, we can deduce that perhaps we are in this startup phase.

**Comment by derekkraan** [comment]
I'm not sure it's possible to tell the difference between lack of quorum because of start up and lack of quorum because there is a network partition. It could be that we are trying to start the application in the presence of a network partition, for example, in which case we will never reach quorum no matter how long we wait (until the network partition is resolved).

I think the correct behaviour is to reject new children while there is no quorum. This can be a pain in the neck, because your system will not operate until quorum has been reached, but I think this is the correct behaviour. To mitigate this, the next thing in your supervision tree should be a process that either links up the parts of the cluster, or ensures that they are linked before continuing with the starting up of any child processes.

I have also been considering the idea that we could enable an `init` callback somehow in `Horde.Supervisor`, perhaps with a "module-based Horde.Supervisor" ala Elixir's Supervisor. This would make it easier to get these things done, and would also ensure that if `Horde.Supervisor` ever crashes, the `init` would always be run again, removing a potential source of bugs from people's systems.

Does this make sense?

**Comment by derekkraan** [comment]
I think that's the best option. The application will not / should not be operational unless we have quorum.

This isn't so much about the internals of Horde, but about the contract that Horde.Supervisor implements, and that should be as close as possible to the one that Elixir's Supervisor itself has. 

From https://hexdocs.pm/elixir/Supervisor.html#start_child/2

> If the child process start function returns {:ok, child} or {:ok, child, info}, then child specification and PID are added to the supervisor and this function returns the same value.

We can't return {:ok, child} if we haven't started the child yet. So in that case we'd have to return something like `{:ok, :deferred}`, but IMO it makes a lot more sense to return `{:error, :quorum_not_met}` and let Horde's users deal with the failure in the way that makes sense for their own applications. In most cases I think this will be just bubbling the error up to higher levels of the application.

**Comment by derekkraan** [comment]
Hi @Ivor ,

I've implemented a behaviour for the distribution strategies. I'm not sure if you're still looking at these issues, but check out #55 if you are. It might clarify some things.

**Comment by Ivor** [comment]
Hi Derek.

Thank you very much. I did implement a static distribution strategy which
works great. It uses a GenServer to keep startup time to skip quorum checks
during the transient startup phase.
I have also built an AWS one which compares the expected nodes with the
ones in the horde to determine quorum. Haven't deployed the AWS one yet.

This seems to make some changes, for example to the members, which might
break my implementation in small wails, so will update to use the behaviour
as soon as I can figure out effects of those changes.

Thank you very much!

**Comment by Ivor** [comment]
Hi Derek.

I have not thought of that since the static one is very simple. The AWS one
is specific to our infernal AWS code so won't be able to do that but could do
the static one.
Might be suitable to just do a blog post with a code sample. I'll let you
know if I do something along those lines.

Thanks again!

**Comment by derekkraan** [comment]
I'm closing this issue since the original issue has been fixed in #55. The discussion can continue below of course.
